### ISSUE TITLE
GetSpecificPortMappingEntry() failed with code 714 (NoSuchEntryInArray)

### ISSUE THREAD

**Issue by Davis-A** [miniupnp/miniupnp#729]
Hello,

I've got miniupnpd running on FreeBSD 14.0, using the pf backend.  I've had some appliances believe they can't initiate upnp.  For example my xbox one.  So onto debugging.

Server details:
```
[05:26] andrew@apu ~> miniupnpd --version
miniupnpd 2.3.3 Apr 10 2024
using pf backend
```

I'm using the FreeBSD pkg which has these [build options](https://www.freshports.org/net/miniupnpd/).
```
The following configuration options are available for miniupnpd-2.3.3_3,1:
     CHECK_PORTINUSE=off: Check if ports are in use
     IPV6=on: IPv6 protocol support
     LEASEFILE=off: Enable lease file
     UPNP_IGDV2=off: Build an IGDv2 instead of an IGDv1
     UPNP_STRICT=off: More strict UPnP specification compliance
```

From a Linux client if i try and open a port using miniupnpc i get the error `GetSpecificPortMappingEntry() failed with code 714 (NoSuchEntryInArray)` and upnpc exits exits with code `2`.

```
[05:29] andrew@neon ~> upnpc -e "upnpc test" -a 10.0.0.10 5001 5001 TCP
upnpc: miniupnpc library test client, version 2.2.7.
 (c) 2005-2024 Thomas Bernard.
More information at https://miniupnp.tuxfamily.org/ or http://miniupnp.free.fr/

List of UPNP devices found on the network :
 desc: http://10.0.0.1:5000/rootDesc.xml
 st: urn:schemas-upnp-org:device:InternetGatewayDevice:1

Found valid IGD : http://10.0.0.1:5000/ctl/IPConn
Local LAN ip address : 10.0.0.10
ExternalIPAddress = 116.255.34.255
GetSpecificPortMappingEntry() failed with code 714 (NoSuchEntryInArray)
```

Client version details here.  
```
[05:29] andrew@neon ~> upnpc --version
upnpc: miniupnpc library test client, version 2.2.7.
 (c) 2005-2024 Thomas Bernard.
More information at https://miniupnp.tuxfamily.org/ or http://miniupnp.free.fr/

List of UPNP devices found on the network :
 desc: http://10.0.0.1:5000/rootDesc.xml
 st: urn:schemas-upnp-org:device:InternetGatewayDevice:1

Found valid IGD : http://10.0.0.1:5000/ctl/IPConn
Local LAN ip address : 10.0.0.10
Unknown switch --
```

But what's more confusing is, if I go query pf then i can see it has added the redirect:

```
[05:32] andrew@apu ~> sudo pfctl -a miniupnpd -s nat
rdr pass quick on igb0 inet proto tcp from any to any port = 5001 keep state label "upnpc test" rtable 0 -> 10.0.0.10 port 5001
```

Any assistance would be appreciated.  I'm more than happy to provide any configuration files, or add a patch with some printf debugging etc.

**Comment by miniupnp** [comment]
> [05:26] andrew@apu ~> miniupnpd --version
> miniupnpd 2.3.3 Apr 10 2024
> using pf backend

please check with latest version which is miniupnpd 2.3.6

**Comment by miniupnp** [comment]
please use ` upnpc -l`  to list redirections from your linux client.

**Comment by Davis-A** [comment]
Well color me surprised!


```
[10:50] andrew@apu ~/d/miniupnpd-2.3.6> ./miniupnpd --version
miniupnpd 2.3.6 Apr 25 2024
using pf backend
build options: ipv6 NAT-PMP PCP
[10:50] andrew@apu ~/d/miniupnpd-2.3.6> sudo ./miniupnpd -d -f /usr/local/etc/miniupnpd.conf
miniupnpd 77346 - - version 2.3.6 starting NAT-PMP/PCP UPnP-IGD ext if igb0 BOOTID=1714042228
miniupnpd 77346 - - HTTP listening on port 5000
miniupnpd 77346 - - Listening for NAT-PMP/PCP traffic on port 5351
miniupnpd 77346 - - level=0 type=20
miniupnpd 77346 - - sdl_index = 2  igb1:0.d.b9.57.69.c9
miniupnpd 77346 - - ST: urn:schemas-upnp-org:device:InternetGatewayDevice:1 (ver=1)
miniupnpd 77346 - - SSDP M-SEARCH from 10.0.0.10:48192 ST: urn:schemas-upnp-org:device:InternetGatewayDevice:1
miniupnpd 77346 - - Single search found
miniupnpd 77346 - - SendSSDPResponse(): 0 bytes to 10.0.0.10:48192 ST: HTTP/1.1 200 OK
CACHE-CONTROL: max-age=1800
ST: urn:schemas-upnp-org:device:InternetGatewayDevice:1
USN: uuid:=28344d00-11db-4c56-b7a3-5b9b7394d525::urn:schemas-upnp-org:device:InternetGatewayDevice:1
EXT:
SERVER: FreeBSD/14.0-RELEASE-p6 UPnP/1.1 MiniUPnPd/2.3.6
LOCATION: http://10.0.0.1:5000/rootDesc.xml
OPT: "http://schemas.upnp.org/upnp/1/0/"; ns=01
01-NLS: 1714042228
BOOTID.UPNP.ORG: 1714042228
CONFIGID.UPNP.ORG: 1337

miniupnpd 77346 - - HTTP REQUEST from 10.0.0.10:37010 : GET /rootDesc.xml (HTTP/1.1)
miniupnpd 77346 - - Host: 10.0.0.1:5000
miniupnpd 77346 - - HTTP REQUEST from 10.0.0.10:37026 : POST /ctl/IPConn (HTTP/1.1)
miniupnpd 77346 - - Host: 10.0.0.1:5000
miniupnpd 77346 - - SOAPAction: urn:schemas-upnp-org:service:WANIPConnection:1#GetStatusInfo
miniupnpd 77346 - - HTTP REQUEST from 10.0.0.10:37036 : POST /ctl/IPConn (HTTP/1.1)
miniupnpd 77346 - - Host: 10.0.0.1:5000
miniupnpd 77346 - - SOAPAction: urn:schemas-upnp-org:service:WANIPConnection:1#GetExternalIPAddress
miniupnpd 77346 - - HTTP REQUEST from 10.0.0.10:37046 : POST /ctl/IPConn (HTTP/1.1)
miniupnpd 77346 - - Host: 10.0.0.1:5000
miniupnpd 77346 - - SOAPAction: urn:schemas-upnp-org:service:WANIPConnection:1#GetExternalIPAddress
miniupnpd 77346 - - HTTP REQUEST from 10.0.0.10:37054 : POST /ctl/IPConn (HTTP/1.1)
miniupnpd 77346 - - Host: 10.0.0.1:5000
miniupnpd 77346 - - SOAPAction: urn:schemas-upnp-org:service:WANIPConnection:1#AddPortMapping
miniupnpd 77346 - - AddPortMapping: ext port 5001 to 10.0.0.10:5001 protocol TCP for: upnpc test leaseduration=0 rhost=
miniupnpd 77346 - - UPnP permission rule 0 matched : port mapping accepted
miniupnpd 77346 - - redirecting port 5001 to 10.0.0.10:5001 protocol TCP for: upnpc test
miniupnpd 77346 - - HTTP REQUEST from 10.0.0.10:37062 : POST /ctl/IPConn (HTTP/1.1)
miniupnpd 77346 - - Host: 10.0.0.1:5000
miniupnpd 77346 - - SOAPAction: urn:schemas-upnp-org:service:WANIPConnection:1#GetSpecificPortMappingEntry
miniupnpd 77346 - - GetSpecificPortMappingEntry: rhost='' 5001 TCP found => 10.0.0.10:5001 desc='upnpc test' duration=0
miniupnpd 77346 - - level=0 type=20
miniupnpd 77346 - - sdl_index = 2  igb1:0.d.b9.57.69.c9
miniupnpd 77346 - - ST: urn:schemas-upnp-org:device:InternetGatewayDevice:1 (ver=1)
miniupnpd 77346 - - SSDP M-SEARCH from 10.0.0.10:57762 ST: urn:schemas-upnp-org:device:InternetGatewayDevice:1
miniupnpd 77346 - - Single search found
miniupnpd 77346 - - SendSSDPResponse(): 0 bytes to 10.0.0.10:57762 ST: HTTP/1.1 200 OK
CACHE-CONTROL: max-age=1800
ST: urn:schemas-upnp-org:device:InternetGatewayDevice:1
USN: uuid:=28344d00-11db-4c56-b7a3-5b9b7394d525::urn:schemas-upnp-org:device:InternetGatewayDevice:1
EXT:
SERVER: FreeBSD/14.0-RELEASE-p6 UPnP/1.1 MiniUPnPd/2.3.6
LOCATION: http://10.0.0.1:5000/rootDesc.xml
OPT: "http://schemas.upnp.org/upnp/1/0/"; ns=01
01-NLS: 1714042228
BOOTID.UPNP.ORG: 1714042228
CONFIGID.UPNP.ORG: 1337

miniupnpd 77346 - - HTTP REQUEST from 10.0.0.10:37066 : GET /rootDesc.xml (HTTP/1.1)
miniupnpd 77346 - - Host: 10.0.0.1:5000
miniupnpd 77346 - - HTTP REQUEST from 10.0.0.10:37068 : POST /ctl/IPConn (HTTP/1.1)
miniupnpd 77346 - - Host: 10.0.0.1:5000
miniupnpd 77346 - - SOAPAction: urn:schemas-upnp-org:service:WANIPConnection:1#GetStatusInfo
miniupnpd 77346 - - HTTP REQUEST from 10.0.0.10:37072 : POST /ctl/IPConn (HTTP/1.1)
miniupnpd 77346 - - Host: 10.0.0.1:5000
miniupnpd 77346 - - SOAPAction: urn:schemas-upnp-org:service:WANIPConnection:1#GetExternalIPAddress
miniupnpd 77346 - - HTTP REQUEST from 10.0.0.10:37086 : POST /ctl/IPConn (HTTP/1.1)
miniupnpd 77346 - - Host: 10.0.0.1:5000
miniupnpd 77346 - - SOAPAction: urn:schemas-upnp-org:service:WANIPConnection:1#GetConnectionTypeInfo
miniupnpd 77346 - - HTTP REQUEST from 10.0.0.10:37098 : POST /ctl/IPConn (HTTP/1.1)
miniupnpd 77346 - - Host: 10.0.0.1:5000
miniupnpd 77346 - - SOAPAction: urn:schemas-upnp-org:service:WANIPConnection:1#GetStatusInfo
miniupnpd 77346 - - HTTP REQUEST from 10.0.0.10:37114 : POST /ctl/CmnIfCfg (HTTP/1.1)
miniupnpd 77346 - - Host: 10.0.0.1:5000
miniupnpd 77346 - - SOAPAction: urn:schemas-upnp-org:service:WANCommonInterfaceConfig:1#GetCommonLinkProperties
miniupnpd 77346 - - HTTP REQUEST from 10.0.0.10:37120 : POST /ctl/IPConn (HTTP/1.1)
miniupnpd 77346 - - Host: 10.0.0.1:5000
miniupnpd 77346 - - SOAPAction: urn:schemas-upnp-org:service:WANIPConnection:1#GetExternalIPAddress
miniupnpd 77346 - - HTTP REQUEST from 10.0.0.10:37128 : POST /ctl/IPConn (HTTP/1.1)
miniupnpd 77346 - - Host: 10.0.0.1:5000
miniupnpd 77346 - - SOAPAction: urn:schemas-upnp-org:service:WANIPConnection:1#GetGenericPortMappingEntry
miniupnpd 77346 - - GetGenericPortMappingEntry: index=0
miniupnpd 77346 - - HTTP REQUEST from 10.0.0.10:37144 : POST /ctl/IPConn (HTTP/1.1)
miniupnpd 77346 - - Host: 10.0.0.1:5000
miniupnpd 77346 - - SOAPAction: urn:schemas-upnp-org:service:WANIPConnection:1#GetGenericPortMappingEntry
miniupnpd 77346 - - GetGenericPortMappingEntry: index=1
miniupnpd 77346 - - Returning UPnPError 713: SpecifiedArrayIndexInvalid
```

And on the client:

```
[10:49] andrew@neon ~> upnpc -e "upnpc test" -a 10.0.0.10 5001 5001 TCP                                                                                                                     upnpc: miniupnpc library test client, version 2.2.7.                                                                                                                                         (c) 2005-2024 Thomas Bernard.                                                                                                                                                              More information at https://miniupnp.tuxfamily.org/ or http://miniupnp.free.fr/
                                                                                                                                                                                            List of UPNP devices found on the network :                                                                                                                                                  desc: http://10.0.0.1:5000/rootDesc.xml                                                                                                                                                     st: urn:schemas-upnp-org:device:InternetGatewayDevice:1
                                                                                                                                                                                            Found valid IGD : http://10.0.0.1:5000/ctl/IPConn                                                                                                                                           Local LAN ip address : 10.0.0.10                                                                                                                                                            ExternalIPAddress = 116.255.34.255
InternalIP:Port = 10.0.0.10:5001
external 116.255.34.255:5001 TCP is redirected to internal 10.0.0.10:5001 (duration=0)
```
```
[10:50] andrew@neon ~> upnpc -l
upnpc: miniupnpc library test client, version 2.2.7.
 (c) 2005-2024 Thomas Bernard.
More information at https://miniupnp.tuxfamily.org/ or http://miniupnp.free.fr/

List of UPNP devices found on the network :
 desc: http://10.0.0.1:5000/rootDesc.xml
 st: urn:schemas-upnp-org:device:InternetGatewayDevice:1

Found valid IGD : http://10.0.0.1:5000/ctl/IPConn
Local LAN ip address : 10.0.0.10
Connection Type : IP_Routed
Status : Connected, uptime=1227426s, LastConnectionError : ERROR_NONE
  Time started : Thu Apr 11 05:53:47 2024
MaxBitRateDown : 8388608 bps (8.3 Mbps)   MaxBitRateUp 4194304 bps (4.1 Mbps)
ExternalIPAddress = 116.255.34.255
 i protocol exPort->inAddr:inPort description remoteHost leaseTime
 0 TCP  5001->10.0.0.10:5001  'upnpc test' '' 0
```

Looks like the client isn't seeing an error anymore.  But, the server logging a `SpecifiedArrayIndexInvalid` error.

Any insights on what in the changelog might have made a difference?

**Comment by miniupnp** [comment]
`pf: fix for OpenBSD 7.4` ?
